### ISSUE TITLE
Exception handling fail on GCC 8.2 (32bit)

### ISSUE THREAD

**Issue by dmitryryzhov** [riscv/riscv-gcc#142]
I have a problem with exception handling during migration of the gcc from 7.2 to 8.2.

This simple test, which throws an exception, works correctly on gcc 7.2, but causes an error on 8.2 (only in 32bit version): 

```
#include <stdio.h>

int main() {
  printf("Warm up\n");
  try {
    throw 1;
  }
  catch (...) {
    printf("In catch\n");
    return 0;
  }
  printf("Back in main\n");
  return 1;
}
```
Compiled with commands (gcc 7.2 similarly):

```
riscv64-unknown-elf-gcc -O2 -g  -mcmodel=medany -Wall -Wextra -Werror -march=rv32imc -mabi=ilp32 -c *.c *.cpp *.S
riscv64-unknown-elf-gcc -march=rv32imc -mabi=ilp32 -Tlink.ld \
        /opt/riscv-gnu-toolchain-elf/bin/../lib/gcc/riscv64-unknown-elf/8.2.0/rv32imc/ilp32/crti.o \
        /opt/riscv-gnu-toolchain-elf/bin/../lib/gcc/riscv64-unknown-elf/8.2.0/rv32imc/ilp32/crtbegin.o \
        *.o -nostartfiles \
        /opt/riscv-gnu-toolchain-elf/bin/../lib/gcc/riscv64-unknown-elf/8.2.0/rv32imc/ilp32/crtend.o \
        /opt/riscv-gnu-toolchain-elf/bin/../lib/gcc/riscv64-unknown-elf/8.2.0/rv32imc/ilp32/crtn.o \
        -lstdc++ -lm -o rv32imc-test.elf
```

An error occurs, when program process .eh_frame section from elf:

```
(gdb) bt
#0  classify_object_over_fdes (ob=ob@entry=0x8001fc64 <object>, this_fde=this_fde@entry=0x8001db04) at ../../../../../../riscv-gcc/libgcc/unwind-dw2-fde.h:174
#1  0x800060a8 in init_object (ob=0x8001fc64 <object>) at ../../../../../../riscv-gcc/libgcc/unwind-dw2-fde.c:777
#2  search_object (ob=ob@entry=0x8001fc64 <object>, pc=pc@entry=0x80004c4b <_Unwind_RaiseException+139>) at ../../../../../../riscv-gcc/libgcc/unwind-dw2-fde.c:989
#3  0x8000680c in _Unwind_Find_FDE (pc=0x80004c4b <_Unwind_RaiseException+139>, bases=bases@entry=0xbffffb08) at ../../../../../../riscv-gcc/libgcc/unwind-dw2-fde.c:1066
#4  0x8000393c in uw_frame_state_for (context=context@entry=0xbffff9f0, fs=fs@entry=0xbffff778) at ../../../../../../riscv-gcc/libgcc/unwind-dw2.c:1257
#5  0x80004620 in uw_init_context_1 (context=context@entry=0xbffff9f0, outer_cfa=outer_cfa@entry=0xbfffffa0, outer_ra=0x80018e44 <__cxa_throw+76>, 
    outer_ra@entry=0x80004c4c <_Unwind_RaiseException+140>) at ../../../../../../riscv-gcc/libgcc/unwind-dw2.c:1586
#6  0x80004c4c in _Unwind_RaiseException (exc=0x80020ad0) at ../../../../../../riscv-gcc/libgcc/unwind.inc:93
#7  0x80018e44 in __cxa_throw ()
#8  0x80017a70 in main () at test.cpp:6
```

According to information from http://refspecs.linuxfoundation.org/LSB_3.0.0/LSB-Core-generic/LSB-Core-generic/ehframechpt.html , the .eh_frame section shall contain 1 or more CFI records, each one contains CIE and 1 or more FDE records.
Generally they start with 4-byte Length field and if it contains the value 0, then this CIE shall be considered as a terminator and processing should stop.

While debugging with gdb everything goes okay about a hundred records (with lengths values of 0x10, 0x20 and others of the same order), until a record with lenght 0x3659bff. It seems that the terminator is missed.

Dumps, produced by readelf, confirms the problem: 
`readelf --debug=frames rv32imc-test.elf > rv32imc-test.eh_dump`

Dump of elf, compiled with gcc 7.2 (zero terminator is present):

```
Contents of the .eh_frame section:

...

000015ec 00000018 000015a8 FDE cie=00000048 pc=80018b30..80018b6c
  DW_CFA_advance_loc: 4 to 80018b34
  DW_CFA_def_cfa_offset: 16
  DW_CFA_advance_loc: 4 to 80018b38
  DW_CFA_offset: r1 at cfa-4
  DW_CFA_advance_loc: 44 to 80018b64
  DW_CFA_restore: r1
  DW_CFA_advance_loc: 4 to 80018b68
  DW_CFA_def_cfa_offset: 0

00001608 ZERO terminator


0000160c 00000010 000015c8 FDE cie=00000048 pc=80013cb0..800140e4
  DW_CFA_nop
  DW_CFA_nop
  DW_CFA_nop

00001620 00000010 000015dc FDE cie=00000048 pc=800140e4..800144f4
  DW_CFA_nop
  DW_CFA_nop
  DW_CFA_nop

Contents of the .debug_frame section:
...
```

But with gcc 8.2 it is missed:

```
Contents of the .eh_frame section:

...

000016c4 00000018 00001680 FDE cie=00000048 pc=80019440..8001947c
  DW_CFA_advance_loc: 4 to 80019444
  DW_CFA_def_cfa_offset: 16
  DW_CFA_advance_loc: 4 to 80019448
  DW_CFA_offset: r1 at cfa-4
  DW_CFA_advance_loc: 44 to 80019474
  DW_CFA_restore: r1
  DW_CFA_advance_loc: 4 to 80019478
  DW_CFA_def_cfa_offset: 0

000016e0 00000010 0000169c FDE cie=00000048 pc=80014644..80014a78
  DW_CFA_nop
  DW_CFA_nop
  DW_CFA_nop

000016f4 00000010 000016b0 FDE cie=00000048 pc=80014a78..80014e88
  DW_CFA_nop
  DW_CFA_nop
  DW_CFA_nop

Contents of the .debug_frame section:
...
```

With gcc 8.2 64-bit (rv64imc), also all is ok:

```
...
00001734 ZERO terminator
...
```

It seems that a problem somewhere in eh_frame section generation, but I can not figure out the exact bug.

**Comment by jim-wilson** [comment]
On Tue, Dec 18, 2018 at 8:26 AM Dmitry Ryzhov <notifications@github.com>
wrote:

> Compiled with commands (gcc 7.2 similarly):
>
> riscv64-unknown-elf-gcc -O2 -g  -mcmodel=medany -Wall -Wextra -Werror -march=rv32imc -mabi=ilp32 -c *.c *.cpp *.S
> riscv64-unknown-elf-gcc -march=rv32imc -mabi=ilp32 -Tlink.ld \
>         /opt/riscv-gnu-toolchain-elf/bin/../lib/gcc/riscv64-unknown-elf/8.2.0/rv32imc/ilp32/crti.o \
>         /opt/riscv-gnu-toolchain-elf/bin/../lib/gcc/riscv64-unknown-elf/8.2.0/rv32imc/ilp32/crtbegin.o \
>         *.o -nostartfiles \
>         /opt/riscv-gnu-toolchain-elf/bin/../lib/gcc/riscv64-unknown-elf/8.2.0/rv32imc/ilp32/crtend.o \
>         /opt/riscv-gnu-toolchain-elf/bin/../lib/gcc/riscv64-unknown-elf/8.2.0/rv32imc/ilp32/crtn.o \
>         -lstdc++ -lm -o rv32imc-test.elf
>
>
The order of object files and libraries on the link line is very
important.  It must be exactly the same as gcc generated or your programs
may not work.  In this case, you have libraries added after crtend.o, but
the unwind zero terminator is in crtend.o, and will not be at the end of
the unwind info unless crtend.o is the very last thing linked in.  If you
look at the gcc generated link line, crtend.o is the very last one on the
link command line.  You are probably getting lucky with other compiler
versions, and getting some zero fill section alignment that is substituting
for the missing zero terminator for the unwind info.

I can't reproduce your problem compiling normally, but I was able to
reproduce your problem by putting crtend.o in the wrong place on the link
line, before the C++ libraries.

Note that for an embedded elf target, crti.o and crtn.o are empty files,
and gcc doesn't bother to include them on the link line.  But for a linux
compiler, it would be crtend.o and then crtn.o at the very end, after the
libraries.

Jim

**Comment by dmitryryzhov** [comment]
Thanks for your reply, it helped me find a solution to the problem.

I actually made a mistake when I was preparing a short example, but in my general script crtend.o-crtn.o pair was the last item in command. But according to your advice, I looked at the gcc generated link line and found that it places lstdc++ before lgcc. And after I swap their groups, the terminator appeared in the elf and the exception test began to work.
It is a little bit magic for me, why line '-lstdc++ -lm -lgcc -lc -lgcc' is worked correctly, but '-lgcc -lc -lgcc -lstdc++ -lm' - is not, and how their sequence affects linker and crtend.
